### PR TITLE
Add Java client for parametric tests

### PR DESCRIPTION
**Pull request by PerfectSlayer**
* Add java tracer to parametric tests.
* Add support for custom binaries to parametric tests (only _java_ client uses it for now).

Java client is not added in matrix for CI as **tests are failing**.
The goal of adding Java client support is to give a way for Java tracer developer to test the metrics coming development.

### CODE REVIEW

**Comment by cbeauchesne** [comment]
Hi @PerfectSlayer , you may need this : https://github.com/DataDog/system-tests/blob/main/docs/edit/lint.md

**Comment by PerfectSlayer** [comment]
Hi @cbeauchesne ! 

Thanks, will do!

Right now, I am thinking about a way to get the parametric test use custom binaries (not the ones pull from release or package manager).
I had a quick chat with @Kyle-Verhoog and we will try to support this feature as all other system tests.

**Comment by cbeauchesne** [comment]
FYI, there is a mechanism for the need, you can read this : https://github.com/DataDog/system-tests/blob/main/docs/execute/binaries.md

**Comment by cbeauchesne** [comment]
I've updated your branch, there were an issue fixed by #519

**Comment by PerfectSlayer** [comment]
Oops sorry, I missed the comment and force pushed my work (I keep doing it until its a draft 😅 ).
I rebased it on `main` branch. I guess it should be fine?

I added the support for custom binaries. I had to change the Docker build context from `parametric/apps/<language>` to `/` in order to access to `/binaires`. (I tried to use symlinks first without success… Docker keeps complaining about it 😞 )
I implemented this features for the Java tracer only, not sure how to do it for the other languages.

**Comment by PerfectSlayer** [comment]
> FYI, there is a mechanism for the need, you can read this : https://github.com/DataDog/system-tests/blob/main/docs/execute/binaries.md

I implemented it in a separate PR: #504 😉

File: parametric/apps/java/src/main/proto/apm_test_client.proto
Comment (Kyle-Verhoog) [inline parametric/apps/java/src/main/proto/apm_test_client.proto:1]: Is there a way we can reuse https://github.com/DataDog/system-tests/blob/main/parametric/protos/apm_test_client.proto?

**Review by Kyle-Verhoog — COMMENTED** [review]
Just one question, otherwise LGTM to merge (even if the tests aren't passing for Java yet).

@PerfectSlayer are we ok to merge soon?

File: parametric/apps/java/src/main/proto/apm_test_client.proto
Comment (PerfectSlayer) [reply parametric/apps/java/src/main/proto/apm_test_client.proto:1]: I think I can configure the maven plugin to use the original file. 
I made a copy because .net client did it too.

**Comment by Kyle-Verhoog** [comment]
> The main issue I have is there is no API in the Java tracer to flush metrics... So it is basically a 15 secs wait to ensure they are flushed (every 10 secs) and it's killing test execution time 😢

Ahh I see. Is the flush interval configurable?

>  Do you know when do you need this feature? So I can ask for the time to finish it.

We don't need to have it enabled right away for the metrics. @ygree is looking to use them for the shared sampling tests that are here. So if we're able to run the tests then I'm good to merge. 😄

**Comment by PerfectSlayer** [comment]
> Ahh I see. Is the flush interval configurable?

I think it is but are the tests relevant if we don't use the default configuration for such type of settings?

> Is there a way we can reuse (the protobuf file)

I struggle to find a way to prevent its duplication.
I can copy it in Docker build image before running Maven build but that means it will only be available for the docker image and local development will be blocked as Maven project won't compile anymore… With is not really great.

I pushed most of my changes after rebasing them on your merge.
 The only test that should be failing is the 0% sampling case (this use case does not seem well defined to fix it or not 🤷 ) 

I still have few changes on my side and I wonder if you are interested by them or not?

- Create a dedicated network for each test suite
- Allow remote debugging for the Java library

@ygree should be able to reuse the Java library integration.

**Comment by Kyle-Verhoog** [comment]
> I think it is but are the tests relevant if we don't use the default configuration for such type of settings?

We could use a delay for just the Java tests for now (better than nothing)


> I struggle to find a way to prevent its duplication.

That's alright. It's not a huge deal, we just have to ensure that the copy stays roughly up to date 🙂 



> I still have few changes on my side and I wonder if you are interested by them or not?

Both of these sound amazing! Can you put the network one up in a separate PR?

The remote debugging one could go in this one.


> What is missing on your side to merge the PR? The proto duplication, anything else?

Nothing further, it's good to me!

**Review by Kyle-Verhoog — APPROVED** [review]
Awesome, thank you @PerfectSlayer!! 🙌

**Comment by PerfectSlayer** [comment]
> We could use a delay for just the Java tests for now (better than nothing). I'm ok to do this in a follow-up PR.

I talk to the team and we will add a proper interface to fix the flushing issue.

> That's alright. It's not a huge deal, we just have to ensure that the copy stays roughly up to date 🙂

Even if the proto is up-to-date in the library, it does not mean the service implementation is 😅 
It might be better to duplicate the protobuf file and expect the new version is compatible with the old one (until it is properly updated) rather than having the Java library compilation fails even if proto are compatible. 

> Both of these sound amazing! Can you put the network one up in a separate PR?

Will do but i would expect the change to be review / updated as I am not a python expert and not sure I did the patch the best way.

**Review by cbeauchesne — APPROVED** [review]

**Pull request merged**
